### DISCUSSION TITLE
Remote Desktop Client for Microsoft outdated

### DISCUSSION THREAD

**Discussion by MSIT-GHM** [microsoft/winget-pkgs#28297] [Q&A] [answered]
Hello,

I've noticed the remote desktop client hasn't been outdated for about 2 months now and would like to update it myself by submitting a manifest. However, the current manifest lists a download URL to that specific version (the old one). The URL I found that points to the updated version (in fact, dynamically speaking, all future versions) is found [here](https://docs.microsoft.com/en-us/azure/virtual-desktop/user-documentation/connect-windows-7-10) by inspecting the link for the x64 client.

1. If I use that link in the manifest, the hash will become outdated the next time Microsoft pushes an update to this program (which is fairly often). There'll be a version mismatch as well.

2. What's the policy here? Create the manifest, use that new link to download the installer, verify the (current) msi package and jot down the hash and package ID and the version, do a pull request and then ride it out until the manifest becomes invalid and I, or someone, uploads an updated manifest?

3. Due to the nature of how quickly Microsoft updates this program, I would prefer if there was an automated solution to this. Are there any guidelines I could follow in maybe exploring automation?

Or, if I'm off my mark entirely, any advice would be welcome. Thank you.

**Comment by jedieaston** [comment, marked as answer]
1. You could use the link that link redirects to, so we can keep older versions of the client in the repo. And then the hash would always be correct for that specific version.

1. You have it right. I don't know how this one was missed.

1. As long as your automation opens a PR, it's fine. If you use that link that always points to the latest version, wingetbot would automatically update it within a few days of the hash mismatch (if someone doesn't catch it sooner. a few days is a long time!)

**Reply by jedieaston** [reply]
Also, didn't mean to steal your thunder, but I opened a PR to get the current version into the repo: https://github.com/microsoft/winget-pkgs/pull/28409.

If you automate it that would be totally appreciated.

**Reply by MSIT-GHM** [reply]
Thanks for the reply and for opening a PR!

Quick followup though: can you elaborate a bit on your reply (number 1)? Specifically on how older versions will be kept if we use a link that redirects to a new version?

**Reply by jedieaston** [reply]
What I often do when we have a URL that redirects to the latest version (as those links are for Remote Desktop) is plug it into something like [WhereGoes](https://wheregoes.com/) (or I have a PowerShell Script I use sometimes) to get the URL that it points to (which often permanently works), and use that in the manifest. So we always have that specific version of the software in the repo. 

For example, the URL in that docs page for the x64 version is https://go.microsoft.com/fwlink/?linkid=2068602, but plugging that into WhereGoes gets me https://query.prod.cms.rt.microsoft.com/cms/api/am/binary/RWJDR5, which is a permalink. Using that permalink in the repo means we always have 1.2.2322.0 even if the latest URL starts pointing elsewhere.

I can pull out that PowerShell snippet if you need it, I use it when getting the direct URLs to the latest version of the Visual Studio Installers.

**Reply by MSIT-GHM** [reply]
OH! That's how you do that! I was wondering where that link was coming from, makes sense now. 

Yeah, using the permalink for version control would be ideal. I was planning on using python for the automation but now that I've learned this bit here, I can incorporate it into a script that submits the PR in the end. I was going to check the link / installer against the current version and go from there if there was a mismatch. 

I'm not promising anything so I don't dissapoint, but if I do were should I post it so others can take over if I stop hosting the script?

**Reply by jedieaston** [reply]
A GitHub Gist, maybe? If you link it here we'll have it around for when we need it.

**Reply by MSIT-GHM** [reply]
Perfect, thank you.

**Reply by vedantmgoyal9** [reply]
@jedieaston please don't mind, but can you tell how many scripts you have? I think you have made scripts for _literally_ everything.

**Reply by jedieaston** [reply]
I probably have four I use for automated updates of things that have decent APIs available (JetBrains, Jackett, Edge, Visual Studio). And I wrote myself a little Discord bot for fun that scrapes GitHub and a couple other sites every few hours to see if there are updates available, which I then just use my duct-taped together PowerShell library to update manually. I write one-off scripts all the time to do bulk metadata changes. 

The PowerShell library was nice, but I'm playing around with rewriting it with something that supports the JSON schemas that are provided in winget-cli so I don't have to manually write out the field names every time I want to do something that isn't an update. I have a prototype working in Python, but there's probably not anything worth making public yet.

**Reply by denelon** [reply]
@BoomEarth take a look at https://github.com/microsoft/winget-create. We have a tool developed to help create manifests. It can even submit the PR.

**Reply by vedantmgoyal9** [reply]
> [...] that have decent APIs available (JetBrains, Jackett, Edge, Visual Studio).

@jedieaston Can you share links to the API documentation or how you use it? I am interested in learning that how they work. Also, if you can share the source code of your Discord Bot, it will be very helpful in making some part of my script?

**Reply by jedieaston** [reply]
I'll try to clean up my scripts when I have a mo, but in the meantime inspiration can be found here: https://github.com/aaronparker/evergreen/. I didn't use this when building mine (which is why they're hacky) but the maintainer over there has built tons of automated infrastructure to update packages (including a sample script to generate winget manifests, although that was pre wingetcreate).
